waiter/src/waiter/token.clj

```diff
@@ -638,60 +639,86 @@
+                                               streaming-timeout-ms)]
+      (let [_ (log/info "request will use streaming timeout of" configured-streaming-timeout-ms "ms")
+            correlation-id (cid/get-correlation-id)
+            watch-chan-xform
+            (comp
+              (map
+                (fn event-filter [{:keys [object type] :as event}]
+                  (cid/cinfo correlation-id "received event from token-watch-maintainer daemon" {:type (:type event)})
+                  (cid/cdebug correlation-id "full tokens event data received from daemon" {:event event})
+                  (case type
+                    :INITIAL
+                    (assoc event :object (->> object
+                                              (filter index-filter-fn)
+                                              (map metadata-transducer-fn)))
+                    :EVENTS
+                    (assoc event :object (->> object
+                                              (filter (fn [{:keys [object]}]
+                                                        (index-filter-fn object)))
+                                              (map (fn [{:keys [type] :as entry}]
+                                                     (if (= type :UPDATE)
+                                                       (update entry :object metadata-transducer-fn)
+                                                       entry)))))
+                    (throw (ex-info "Invalid event type provided" {:event event})))))
+              (filter
+                (fn empty-aggregate-events [{:keys [object type] :as event}]
+                  (case type
+                    :INITIAL true
+                    :EVENTS (not-empty object)
+                    (throw (ex-info "Invalid event type provided" {:event event})))))
+              (map
+                (fn [event]
+                  (cid/cinfo correlation-id "forwarding tokens event to client" {:type (:type event)})
+                  (cid/cdebug correlation-id "full tokens event data being sent to client" {:event event})
+                  (utils/clj->json event))))
+            watch-chan-ex-handler-fn
+            (fn watch-chan-ex-handler [e]
+              (async/put! ctrl e)
+              (cid/cerror correlation-id e "error during transformation of a token watch event"))
+            watch-chan-buffer (async/buffer 1000)
+            watch-chan (async/chan watch-chan-buffer watch-chan-xform watch-chan-ex-handler-fn)
+            _ (log/info "created watch-chan" watch-chan)
+            watch-mult (async/mult watch-chan)
+            response-chan (async/chan 1024)
+            _ (async/tap watch-mult response-chan)
+            trigger-chan (au/latest-chan)
+            _ (async/tap watch-mult trigger-chan)]
+        (async/go-loop [timeout-ch (async/timeout configured-streaming-timeout-ms)]
+          (let [[message chan] (async/alts! [trigger-chan timeout-ch] :priority true)]
```

Comment:
We can avoid the mult and extra channel if we read off the watch-chan directly and push to the response-chan with the message, and do cleanup if the timeout-ch triggers or if the response-chan errors when calling put! on it.

Reply:
I wanted to avoid side-effects being triggered from the map/reduce functions.